### PR TITLE
8241390: 'Deadlock' with VM_RedefineClasses::lock_classes()

### PR DESCRIPTION
**Pull request by datior**
The change fixes a 'deadlock' situation in VM_RedefineClasses::lock_classes() when the current thread is in the middle of redefining the same class. The change is based on the fix [1] suggested in the Jira issue [2] comment.

[1] http://cr.openjdk.java.net/~jiangli/8241390/webrev.00/
[2] https://bugs.openjdk.java.net/browse/JDK-8241390 

Testing: :jdk_instrument, tier1-tier3, and tier5 tests pass.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8241390](https://bugs.openjdk.java.net/browse/JDK-8241390): 'Deadlock' with VM_RedefineClasses::lock_classes()

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/190/head:pull/190`
`$ git checkout pull/190`

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp [hunk 1]
```diff
@@ -96,31 +96,67 @@ static inline InstanceKlass* get_ik(jclass def) {
...
+      assert(def_ik == k, "unlocking wrong class");
+    }
+    assert(def_ik->is_being_redefined(),
            "should be being redefined to get here");
-    get_ik(_class_defs[i].klass)->set_is_being_redefined(false);
+
+    // Unlock after we finish all redefines for this class within the thread.
+    if (!cls->contains(def_ik)) {
+      def_ik->set_is_being_redefined(false);
+    }
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiRedefineClasses.cpp:159]: Ok, so adding the Klass to the thread local list for each recursion works like ref counting.  Can't think of a simpler way to do this.  Looks good.

**Review by coleenp — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp (hunk 1 above)
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiRedefineClasses.cpp:159]: Yes, the same class can be pushed to the list multiple times (not more than once by each recursive redefinition). It'd make sense to add a comment about it as it is not obvious.

File: test/jdk/java/lang/instrument/MakeAgentJAR.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/jdk/java/lang/instrument/MakeAgentJAR.sh:1]: I'm second for this suggestion from Coleen to get rid of the shell script in the test.

**Comment by sspitsyn** [comment]
The identifier 'cls' does not reflect the nature of the object and needs to be replaces with something like:
  classes, redef_classes or redef_classes_list
This impacts the files: jvmtiThreadState.hpp and jvmtiRedefineClasses.cpp.

**Review by sspitsyn — CHANGES_REQUESTED** [review]
The following test @summary needs some tweaks:
 * @summary Test recursively redefines the same class. The test hangs if
 * a deadlock happens.
Please, replace: redefines => retransforms.
Also, it'd be nice to align "a deadlock happens" with the "Test recursively ...".

**Review by coleenp — APPROVED** [review]
Nice!  Thank you for fixing this.

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**